src/firmware_new/boards/frankie_v1/io/proto_multicast_communication.c

```diff
@@ -0,0 +1,242 @@
+        (ProtoMulticastCommunicationProfile_t*)communication_profile;
+
+    osEventFlagsWait(networking_event, NETIF_CONFIGURED, osFlagsWaitAny, osWaitForever);
+
+    // Bind the socket to the multicast address and port we then use that
+    // communication profile to join the specified multicast group.
+    //
+    // We use IP6_ADDR_ANY which will default to the ethernet interface on the STM32H7
+    struct netconn* conn = netconn_new(NETCONN_UDP_IPV6);
+    netconn_set_ipv6only(conn, true);
+
+    netconn_bind(conn, IP6_ADDR_ANY, sender_profile->port);
+    netconn_join_leave_group(conn, &sender_profile->multicast_address, IP6_ADDR_ANY,
+                             NETCONN_JOIN);
+
+    // this buffer is used to hold serialized proto
+    uint8_t buffer[sender_profile->message_max_size];
+
+    // network buffer
+    struct netbuf* tx_buf = netbuf_new();
+
+    for (;;)
+    {
+        pb_ostream_t stream = pb_ostream_from_buffer(buffer, sizeof(buffer));
+
+        // block until this task receives a signal that the protobuf has been updated
+        io_proto_multicast_communication_profile_blockUntilEvents(sender_profile,
+                                                                  PROTO_UPDATED);
+
+        // serialize proto into buffer
+        // we ingore the error returned by pb_encode, it is up to the receiver to handle
+        // malformed proto, so we send the buffer regardless
+        io_proto_multicast_communication_profile_acquireLock(sender_profile);
+        pb_encode(&stream, sender_profile->message_fields,
+                  sender_profile->protobuf_struct);
+        io_proto_multicast_communication_profile_releaseLock(sender_profile);
+
+        netbuf_alloc(tx_buf, stream.bytes_written);
+
+        // package payload and send over udp
+        tx_buf->p->payload = buffer;
+        netconn_sendto(conn, tx_buf, &sender_profile->multicast_address,
+                       sender_profile->port);
+    }
+
+    // we should never get here
+    netbuf_delete(tx_buf);
+    netconn_delete(conn);
+}
+
+void io_proto_multicast_listener_task(void* communication_profile)
+{
+    ProtoMulticastCommunicationProfile_t* listener_profile =
+        (ProtoMulticastCommunicationProfile_t*)communication_profile;
+
+    osEventFlagsWait(networking_event, NETIF_CONFIGURED, osFlagsWaitAny, osWaitForever);
+
+    // Bind the socket to the multicast address and port we then use that
+    // communication profile to join the specified multicast group.
+    struct netconn* conn = netconn_new(NETCONN_UDP_IPV6);
+    netconn_set_ipv6only(conn, true);
+    netconn_set_recvtimeout(conn, NETWORK_TIMEOUT_MS);
+
+    netconn_bind(conn, &listener_profile->multicast_address, listener_profile->port);
+    netconn_join_leave_group(conn, &listener_profile->multicast_address, IP6_ADDR_ANY,
+                             NETCONN_JOIN);
+
+    struct netbuf* rx_buf = NULL;
+    err_t network_err;
+    int no_protobuf_err;
```

Comment:
double checking this should be an int and not a bool? (assuming nanopb may return multiple error codes rather than a generic "failure")

Reply:
good point, changed to bool